### PR TITLE
new version of seq2seq

### PR DESCRIPTION
**Pull request by alexholdenmiller**
- primary change is massive updates to further modularize and clean out modules.py

- successfully trains & evals & ranks on babi:task10k:1
- successfully trains & evals & ranks on convai2

### CODE REVIEW

**Comment by alexholdenmiller** [comment]
trains / evals successfully on convai2, including ranking

File: parlai/agents/seq2seq/seq2seq.py [hunk 1]
```diff
@@ -38,17 +38,17 @@ class Seq2seqAgent(TorchAgent):
...
 
-    @staticmethod
-    def add_cmdline_args(argparser):
+    @classmethod
+    def add_cmdline_args(cls, argparser):
         """Add command-line arguments specifically for this agent."""
         agent = argparser.add_argument_group('Seq2Seq Arguments')
         agent.add_argument('--init-model', type=str, default=None,
-                           help='load dict/features/weights/opts from this file')
+                           help='load dict/model/opts from this path')
```
Comment (uralik) [inline parlai/agents/seq2seq/seq2seq.py]: I really like the way when we have automatic default in the help with `%(default)s`, prolly we can fix that in seq2seq now

Comment (alexholdenmiller) [reply parlai/agents/seq2seq/seq2seq.py]: yes let's do separate diff for those, I also want to reformat the whitespace on all the args

File: parlai/agents/seq2seq/modules.py [hunk 2]
```diff
@@ -5,17 +5,34 @@
...
+
+    :param tensor: vector to pad
+    :param length: new length
+    :param dim: (default 0) dimension to pad
+
+    :returns: padded tensor if the tensor is shorter than length
+    """
     if tensor.size(dim) < length:
         return torch.cat(
             [tensor, tensor.new(*tensor.size()[:dim],
```
Comment (stephenroller) [inline parlai/agents/seq2seq/modules.py]: Just out of pure curiosity, did you benchmark this with other approaches? (namely init the full zero matrix, then assign the tensor into it?). Not that there's need to.

Comment (stephenroller) [inline parlai/agents/seq2seq/modules.py]: Just out of pure curiosity, did you benchmark this with other approaches? (namely init the full zero matrix, then assign the tensor into it?). Not that there's need to.

Comment (stephenroller) [reply parlai/agents/seq2seq/modules.py]: Also, should this actually fill_(null_idx)?

File: parlai/agents/seq2seq/modules.py [hunk 3]
```diff
@@ -27,277 +44,291 @@ def pad(tensor, length, dim=0):
...
-        xs -- input to the encoder
-        ys -- expected output from the decoder
-        cands -- set of candidates to rank, if applicable
-        cand_indices -- indices to match candidates with their appropriate xs
-        prev_enc -- if you know you'll pass in the same xs multiple times and
-            the model is in eval mode, you can pass in the encoder output from
-            the last forward pass to skip recalcuating the same encoder output
-        rank_during_training -- (default False) if set, ranks any available
-            cands during training as well
+        :param xs:          (bsz x seqlen) LongTensor input to the encoder
```
Comment (stephenroller) [inline parlai/agents/seq2seq/modules.py]: <3

File: parlai/agents/seq2seq/modules.py [hunk 4]
```diff
@@ -27,277 +44,291 @@ def pad(tensor, length, dim=0):
...
+            return prev_enc
+        else:
+            return self.encoder(xs)
+
+    def _starts(self, bsz):
+        """Return bsz start tokens."""
+        return self.START.detach().expand(bsz, 1)
+
+    def _decode_forced(self, ys, encoder_states):
+        """Decode with teacher forcing."""
```
Comment (stephenroller) [inline parlai/agents/seq2seq/modules.py]: I don't understand what this means

File: parlai/agents/seq2seq/modules.py [hunk 5]
```diff
@@ -27,277 +44,291 @@ def pad(tensor, length, dim=0):
...
-        if isinstance(scores, list):
-            scores = torch.cat(scores, 1)
-
-        return predictions, scores, cand_preds, cand_scores, encoder_states, nbest_beam_preds, nbest_beam_scores
-
-
-class Encoder(nn.Module):
-    def __init__(self, num_features, padding_idx=0, rnn_class='lstm',
-                 emb_size=128, hidden_size=128, num_layers=2, dropout=0.1,
+            scores = self._decode(encoder_states, maxlen or self.longest_label)
```
Comment (stephenroller) [inline parlai/agents/seq2seq/modules.py]: All much cleaner now, great!

File: parlai/agents/seq2seq/modules.py [hunk 6]
```diff
@@ -27,277 +44,291 @@ def pad(tensor, length, dim=0):
...
+
+    def __init__(self, num_features, embeddingsize, hiddensize,
+                 padding_idx=0, rnn_class='lstm', numlayers=2, dropout=0.1,
                  bidirectional=False, shared_lt=None, shared_rnn=None,
                  sparse=False):
+        """Initialize recurrent encoder."""
         super().__init__()
 
         self.dropout = nn.Dropout(p=dropout)
-        self.layers = num_layers
```
Comment (stephenroller) [inline parlai/agents/seq2seq/modules.py]: Why change all the names?

File: parlai/agents/seq2seq/modules.py [hunk 7]
```diff
@@ -309,278 +340,199 @@ def forward(self, xs):
...
-                hidden = (hidden[0].view(-1, self.dirs, bsz, self.hsz).max(1)[0],
-                          hidden[1].view(-1, self.dirs, bsz, self.hsz).max(1)[0])
+                hidden = (hidden[0].view(-1, self.dirs, bsz, self.hsz).sum(1),
+                          hidden[1].view(-1, self.dirs, bsz, self.hsz).sum(1))
             else:
-                hidden = hidden.view(-1, self.dirs, bsz, self.hsz).max(1)[0]
+                hidden = hidden.view(-1, self.dirs, bsz, self.hsz).sum(1)
 
-        return encoder_output, hidden
+        return hidden, encoder_output, attn_mask
```
Comment (stephenroller) [inline parlai/agents/seq2seq/modules.py]: intentional swap?

File: parlai/agents/seq2seq/modules.py [hunk 8]
```diff
@@ -599,49 +551,81 @@ def __init__(self, attn_type, hidden_size, emb_size, bidirectional=False,
...
-                    torch.bmm(hid, enc_out.transpose(1, 2)).squeeze(1))
-            # calculate activation scores
+                enc_t = enc_out.transpose(1, 2)
+                attn_w_premask = torch.bmm(hid, enc_t).squeeze(1)
+
+            # calculate activation scores, apply mask if needed
             if attn_mask is not None:
                 # remove activation from NULL symbols
+                # TODO: is this the best operation?
                 attn_w_premask -= (1 - attn_mask) * 1e20
```
Comment (stephenroller) [inline parlai/agents/seq2seq/modules.py]: Could do `th.masked_fill_(attn_w_premask, -1e20)`. That's what I've seen in fairseq.

Maybe make that `1e20` as a private constant, like `_ALMOST_INFINITY`? Don't think it matters.

File: parlai/agents/seq2seq/seq2seq.py [hunk 9]
```diff
@@ -312,12 +312,12 @@ def train_step(self, batch):
...
         self.model.train()
         self.zero_grad()
-        preds = None
         try:
             out = self.model(batch.text_vec, batch.label_vec)
 
             # generated response
-            preds, scores = out[0], out[1]
+            scores = out[0]
+            preds = scores.max(2)[1]
```
Comment (stephenroller) [inline parlai/agents/seq2seq/seq2seq.py]: nit: i prefer `_, preds = scores.max(2)`

File: parlai/agents/seq2seq/seq2seq.py [hunk 10]
```diff
@@ -382,12 +381,13 @@ def eval_step(self, batch):
...
-        if cand_preds is not None:
+        if cand_scores is not None:
+            cand_preds = cand_scores.sort(1, True)[1]
             # now select the text of the cands based on their scores
-            cand_choices = self._pick_cands(cand_preds, cand_inds,
+            cand_choices = self._pick_cands(cand_preds, cand_params[1],
                                             batch.candidates)
 
-        text = [self._v2t(p) for p in preds]
+        text = [self._v2t(p) for p in scores.max(2)[1].cpu()]
```
Comment (stephenroller) [inline parlai/agents/seq2seq/seq2seq.py]: IMHO hiding this max() call in here is a bit nefarious.

```python
_, preds = scores.max(2)
text = [self._v2t(p) for p in preds.cpu()]
```

File: parlai/core/torch_agent.py [hunk 11]
```diff
@@ -775,7 +775,7 @@ def _init_cuda_buffer(self, model, criterion, batchsize, maxlen):
             try:
                 print('preinitializing pytorch cuda buffer')
                 dummy = torch.ones(batchsize, maxlen).long().cuda()
-                sc = model(dummy, dummy)[1]
+                sc = model(dummy, dummy)[0]
```
Comment (stephenroller) [inline parlai/core/torch_agent.py]: this needs an explanation. It's assuming model has a particular interface and that interface isn't clear to me right now.

File: parlai/agents/seq2seq/modules.py (hunk 7 above)
Comment (stephenroller) [reply parlai/agents/seq2seq/modules.py]: nit TBH I think encoder_output, hidden, attn_mask is better.

File: parlai/core/torch_agent.py (hunk 11 above)
Comment (stephenroller) [reply parlai/core/torch_agent.py]: (Assuming this is related to the hidden, encoder_out, attn_mask from earlier)

File: parlai/agents/seq2seq/modules.py (hunk 4 above)
Comment (uralik) [reply parlai/agents/seq2seq/modules.py]: _decode_forced is putting tokens from the target as the next input to estimate the target score during training or during eval PPL computation

**Comment by alexholdenmiller** [comment]
Note: I squashed old pretrained seq2seq module weights into the new modules and got exactly the same f1, ppl, hits on convai2.

File: parlai/agents/seq2seq/modules.py (hunk 2 above)
Comment (alexholdenmiller) [reply parlai/agents/seq2seq/modules.py]: didn't benchmark, although 1) this is copied from pytorch source, with `dim` param added by me and 2) only used for candidate score padding so I'm not super concerned about the perf :)  oh I'll update with null_idx

File: parlai/agents/seq2seq/modules.py (hunk 6 above)
Comment (alexholdenmiller) [reply parlai/agents/seq2seq/modules.py]: just for consistency, there was a mix of underscores and non and things didn't match with the opt args

File: parlai/agents/seq2seq/modules.py (hunk 7 above)
Comment (alexholdenmiller) [reply parlai/agents/seq2seq/modules.py]: swap was intentional because later
```
hidden = encoder_states[0]
attn_params = (encoder_states[1], encoder_states[2])
```
but I could change that to [1] and ([0], [2])

it does match the rnn output more closely

File: parlai/agents/seq2seq/modules.py (hunk 8 above)
Comment (alexholdenmiller) [reply parlai/agents/seq2seq/modules.py]: `self._YUGE`?

File: parlai/agents/seq2seq/seq2seq.py (hunk 10 above)
Comment (alexholdenmiller) [reply parlai/agents/seq2seq/seq2seq.py]: actually introduced a bug too because I used the teacher-forced scores on accident from the if block, I'll update

File: parlai/core/torch_agent.py (hunk 11 above)
Comment (alexholdenmiller) [reply parlai/core/torch_agent.py]: agreed, I need to find a better way to extract the scores here or move it back to s2s

File: parlai/agents/seq2seq/modules.py (hunk 4 above)
Comment (alexholdenmiller) [reply parlai/agents/seq2seq/modules.py]: "teacher-forcing" always inputs the true token at each time step, vs using the model's predicted
